src/Algebra/Properties/Semiring/Binomial.agda

```diff
@@ -0,0 +1,289 @@
+        k   = toℕ i
+        n-k = n ∸ k
+
+      binomial : Carrier
+      binomial = (x ^ k) * (y ^ n-k)
+
+      term : Carrier
+      term = (n C k) × binomial
+
+    expansion : Carrier
+    expansion = ∑[ i < suc n ] term i
+
+    term₁ : (i : Fin (suc (suc n))) → Carrier
+    term₁ zero    = 0#
+    term₁ (suc i) = x * term i
+
+    sum₁ : Carrier
+    sum₁ = ∑[ i < suc (suc n) ] term₁ i
+
+    term₂ : (i : Fin (suc (suc n))) → Carrier
+    term₂ i with view i
+    ... | top   = 0#
+    ... | inj j = y * term j
+
+    sum₂ : Carrier
+    sum₂ = ∑[ i < suc (suc n) ] term₂ i
+
+------------------------------------------------------------------------
+-- Properties
+
+  module BinomialLemmas (n : ℕ) where
+
+    open Binomial n
+
+    open ≈-Reasoning
+
+    lemma₁ : x * expansion ≈ sum₁
+    lemma₁ = begin
+      x * expansion
+        ≈⟨ *-distribˡ-sum x term ⟩
+      ∑[ i < suc n ] (x * term i)
+        ≈˘⟨ +-identityˡ _ ⟩
+      (0# + ∑[ i < suc n ] (x * term i))
+        ≡⟨⟩
+      (0# + ∑[ i < suc n ] term₁ (suc i))
+        ≡⟨⟩
+      sum₁  ∎
+
+    lemma₂ : y * expansion ≈ sum₂
+    lemma₂ = begin
+      (y * expansion)
+        ≈⟨ *-distribˡ-sum y term ⟩
+      ∑[ i < suc n ] (y * term i)
+        ≈˘⟨ +-identityʳ _ ⟩
+      ∑[ i < suc n ] (y * term i) + 0#
+        ≈⟨ +-cong (sum-cong-≋ lemma₂-inj) lemma₂-top ⟩
+      (∑[ i < suc n ] term₂-inj i) + term₂ [n+1]
+        ≈˘⟨ sum-init-last term₂ ⟩
+      sum term₂
+        ≡⟨⟩
+      sum₂  ∎
+      where
+        [n+1] = fromℕ (suc n)
+
+        lemma₂-top : 0# ≈ term₂ [n+1]
+        lemma₂-top rewrite view-top (suc n) = refl
+
+        term₂-inj : (i : Fin (suc n)) → Carrier
+        term₂-inj i = term₂ (inject₁ i)
+
+        lemma₂-inj : ∀ i → y * term i ≈ term₂-inj i
+        lemma₂-inj i rewrite view-inj i = refl
+
+------------------------------------------------------------------------
+-- Next, a lemma which is independent of commutativity
+
+  module _ {n} (i : Fin (suc n)) where
+
+    open Binomial n using (term)
+
+    x*lemma : x * term i ≈ (n C toℕ i) × Binomial.binomial (suc n) (suc i)
+    x*lemma = begin
+      x * term i                                  ≡⟨⟩
+      x * ((n C k) × (x ^ k * y ^ (n ∸ k)))       ≈˘⟨ *-congˡ (×-assoc-* (n C k) _ _) ⟩
+      x * ((n C k) × x ^ k * y ^ (n ∸ k))         ≈˘⟨ *-assoc x ((n C k) × x ^ k) _ ⟩
+      (x * (n C k) × x ^ k) * y ^ (n ∸ k)         ≈⟨ *-congʳ (×-comm-* (n C k) _ _) ⟩
+      (n C k) × x ^ (suc k) * y ^ (n ∸ k)         ≡⟨⟩
+      (n C k) × x ^ (suc k) * y ^ (suc n ∸ suc k) ≈⟨ ×-assoc-* (n C k) _ _ ⟩
+      (n C k) × Binomial.binomial (suc n) (suc i) ∎
+      where
+        open ≈-Reasoning
+        k = toℕ i
+
+
+------------------------------------------------------------------------
+-- Next, a lemma which does require commutativity
+
+  module _ (x*y≈y*x : x * y ≈ y * x) where
+
+    module _ (n : ℕ) where
+
+      open Binomial n using (binomial; term; term₁; term₂)
+
+      y*lemma : (j : Fin n) → let i = inject₁ j in
```

Comment:
Why for `x*lemma` do you have one module taking `n` and `j` simultaneously, but for `y*lemma` you have a module parameterized on `n` with then `j` as an argument to the lemma? If `j` was a module parameter you could define `i` as a local private definition in the module